docs/design-documents/hal/0004-pin-names-general-guidelines.md

```diff
@@ -0,0 +1,156 @@
+# General Pin Names design document
+
+# Table of contents
+
+1. [Table of contents](#table-of-contents).
+    1. [Revision history](#revision-history).
+1. [Introduction](#introduction).
+    1. [Overview and assumptions](#overview-and-assumptions)
+1. [Detailed design](#detailed-design).
+    1. [LEDs](#leds).
+    1. [Buttons](#buttons).
+1. [Non-valid definitions](#non-valid-definitions).
+1. [Testing compliance](#testing-compliance).
+
+
+### Revision history
+
+Initial revision: 1.0
+
+Authors: Malavika Sajikumar (ADI), Sean Doyle (ADI), Kyle Jansen (ADI), Jerome Coutant (ST), Marcelo Salazar (Arm).
+
+This document is written for Mbed OS 6.
+
+## Introduction
+
+### Overview and assumptions
+
+Mbed OS is designed so that application code written in the platform is portable across different Mbed supported boards with the same hardware capabilities or interfaces. However, the code may not be truly portable due to the differences in pin name definitions for the same kind of interfaces across different boards. 
+
+This document provides general guidelines on best practices for defining pin names that could apply to all boards but it's not specific to any type of connector.
+
+Note there might be separate documents for pin names that apply to specific connectors such as the Arduino Uno. These are available in the [HAL design documents](./) folder.
+
+## Detailed design
+
+To achieve meaningful portability of application code across various Mbed Enabled boards, certain pin names of commonly used interfaces and board components should be common across these boards. This document describes a set of rules on how to define components found on actual boards, using the PinNames.h file as part of the BSP of the board.
+
+### LEDs
+
+**Definition of LEDs**
+
+Only add LEDs that are available on the board. This is an example on how to define LEDs in PinNames.h:
+
+    // Px_xx relates to the processor pin connected to the LED
+    #define LED1 = Px_xx  // LED1
```

Comment:
Most of the platforms define `typedef enum { LED0 = Pin1; .... }  PinName`

This document is changing this to use preprocessor , shouldn't it be explained ? If I read this document and look at PinNames in our tree, it is not the same - a reason for switching to preprocessor should be stated as well..

One maybe related but probably for another time:  why not using enum class instead of plain enum or macros (or combination that would allow us to check if enum is valid and defined) ? See https://www.modernescpp.com/index.php/c-core-guidelines-rules-for-enumerations

Reply:
I've updated the doc to clarify that LEDs and buttons move to defines and affects the preprocesor.